text/0083-rawkv-change-data-capture.md

```diff
@@ -0,0 +1,152 @@
+
+**Checkpoint** will be affected by **Garbage Collection**. **Checkpoint** requires that all versions since **Checkpoint** are kept, otherwise **TiKV-CDC** would miss some changes. So **TiKV-CDC** will set a **Service Safe Point** on timestamp the same as **Checkpoint**, with **TTL** defaults to `24` hours. The **TTL** means that **TiKV-CDC** and replicate tasks should be resumed in no more than **TTL**, otherwise the **Checkpoint** will be invalid.
+
+Finally, in this proposal, one replication task will be processed in only one **Capture**, to make it right more easily. Splitting a task to sub-tasks will get better performance and scalability, but we must be caution to avoid any overlap or hole among sub-tasks, otherwise the correctness will be violated. Sub-tasks will be considered later.
+
+### Recovery Point Objective
+
+Customers require that [Recovery Point Objective] (abbr. **RPO**) should be no more than tens of seconds. In **RawKV CDC**, [RPO] is measured by the **Checkpoint Lag** metric, the interval between **Checkpoint** and **TSO** at the time.
+
+Main factors affecting **Checkpoint Lag** includes:
+
+- TiKV internal:
+  - **Resolved-ts Lag**: the interval between **resolved-ts** and **TSO**. Which approximately equals to the longest duration of all RawKV writes at that moment. In most circumstances, it should be no more than several seconds.
+- Data transmission on network:
+  - **Network latency** cross cities would be `30` milliseconds or even more. But as we are using the high-performance [gRPC] as underlying implementation of networking, and writing to downstream by batch, we don't concern with network latency now.
+- **TiKV-CDC** internal:
+  - **Sorter** would be the bottleneck, as it 1) process data from all regions, 2) holds too many data if **resolved-ts** or downstream is blocked, 3) sorts data in disks when data is too big to fit in memory.
+  - This issue will be even more severe on initial replication. In this scenario **Sorer** must hold all existing data from starting point of replication task to current before send to downstream.
+  - To address this issue, we provide an alternative: using [TiKV-BR] to complete the initial replication. [TiKV-BR] backup and restore data fully utilizing power of the whole TiKV cluster, and don't need the sorting, so is much more efficient.
+
+## Detailed Design
+
+### TiKV
+
+#### Resolved-ts
+
+We will use [memory locks] to trace the "inflight" RawKV writes: acquire a lock before scheduling the writes, and release the lock after (asynchronously) send response to client.
+
+Then, CDC worker in TiKV is able to get the **resolved-ts** as `min(timestamp)-1` of all locks. If there is no lock, use current TSO as **resolved-ts**.
```

Comment:
As concurrency manager [lock the ts before key write](https://github.com/tikv/tikv/blob/a80ab9880d6c25fe9b962e582846360dd229783b/src/storage/mod.rs#L1921-L1925). The `resolved-ts` use the lock ts is enough, no need `-1`.